File: JasSwift/Sources/Tools/TypeTool.swift
import AppKit
import Foundation

// MARK: - Type tool with native in-place text editing
//
// Click on existing unlocked text to edit; click on empty canvas to
// create a new empty Text element and enter editing immediately. Drag
// on empty canvas to create an area-text element. While editing, mouse
// drag extends the selection and standard editing keys flow through
// `onKeyEvent`.

private func nowMs() -> Double { textEditNowMs() }
private func cursorVisible(_ epochMs: Double) -> Bool { textEditCursorVisible(epochMs) }

private struct TextRender {
    let x: Double
    let y: Double
    let fontSize: Double
    let textWidth: Double
    let textHeight: Double
    let fill: Fill?
    let stroke: Stroke?
    let content: String
}

/// Editing-overlay bounds for a Text element. `(x, y)` is the *top* of
/// the layout box (the baseline is `y + 0.8 * fontSize`); for area text
/// the bounds come from `(width, height)` directly, for point text from
/// the rendered line widths and the line count.
private func textDrawBounds(_ elem: Element) -> (Double, Double, Double, Double) {
    guard case .text(let t) = elem else { return (0, 0, 0, 0) }
    if t.width > 0 && t.height > 0 {
        return (t.x, t.y, max(t.width, 1), max(t.height, 1))
    }
    let raw = t.content.isEmpty ? " " : t.content
    let lines = raw.split(separator: "\n", omittingEmptySubsequences: false)
    var maxW: Double = 0
    for l in lines {
        let w = renderedTextWidth(String(l), family: t.fontFamily,
                                  weight: t.fontWeight, style: t.fontStyle, size: t.fontSize)
        if w > maxW { maxW = w }
    }
    let h = Double(lines.count) * t.fontSize
    return (t.x, t.y, max(maxW, 1), h)
}

private func inBox(_ b: (Double, Double, Double, Double), _ x: Double, _ y: Double) -> Bool {
    x >= b.0 && x <= b.0 + b.2 && y >= b.1 && y <= b.1 + b.3
}

/// Hit-test all unlocked Text elements in the document, top-down.
private func hitTestText(_ doc: Document, _ x: Double, _ y: Double) -> (ElementPath, Element)? {
    var result: (ElementPath, Element)?
    func walk(_ elem: Element, _ path: ElementPath) {
        if result != nil { return }
        switch elem {
        case .layer(let l):
            for (i, c) in l.children.enumerated() { walk(c, path + [i]) }
        case .group(let g):
            if !g.locked {
                for (i, c) in g.children.enumerated() { walk(c, path + [i]) }
            }
        case .text:
            if !elem.isLocked && inBox(textDrawBounds(elem), x, y) {
                result = (path, elem)
            }
        default:
            break
        }
    }
    for (li, layer) in doc.layers.enumerated() {
        walk(.layer(layer), [li])
    }
    return result
}

/// Drag-to-create state machine. Sealed so "dragging without a start"
/// is unrepresentable.
private enum DragState {
    case idle
    case dragging(startX: Double, startY: Double, curX: Double, curY: Double)
}

class TypeTool: CanvasTool {
    private var state: DragState = .idle
    private var session: TextEditSession?
    private var didSnapshot = false
    private var hoverText = false

    /// Public test accessor.
    var currentSession: TextEditSession? { session }

    private func buildLayout(_ ctx: ToolContext) -> (TextRender, TextLayout)? {
        guard let s = session, s.target == .text else { return nil }
        guard pathIsValid(ctx.document, s.path) else { return nil }
        guard case .text(let t) = ctx.document.getElement(s.path) else { return nil }
        let measure = makeMeasurer(family: t.fontFamily, weight: t.fontWeight,
                                   style: t.fontStyle, size: t.fontSize)
        let maxW = (t.width > 0 && t.height > 0) ? t.width : 0.0
        let lay = layoutText(s.content, maxWidth: maxW, fontSize: t.fontSize, measure: measure)
        let tr = TextRender(x: t.x, y: t.y, fontSize: t.fontSize,
                            textWidth: t.width, textHeight: t.height,
                            fill: t.fill, stroke: t.stroke, content: s.content)
        return (tr, lay)
    }

    private func cursorAt(_ ctx: ToolContext, _ x: Double, _ y: Double) -> Int {
        guard let (tr, lay) = buildLayout(ctx) else { return 0 }
        return lay.hitTest(x - tr.x, y - tr.y)
    }

    private func ensureSnapshot(_ ctx: ToolContext) {
        if !didSnapshot {
            ctx.snapshot()
            didSnapshot = true
        }
    }

    private func syncToModel(_ ctx: ToolContext) {
        guard let s = session else { return }
        if let newDoc = s.applyToDocument(ctx.document) {
            ctx.controller.setDocument(newDoc)
        }
    }

    private func beginSessionExisting(_ ctx: ToolContext, path: ElementPath, elem: Element, cursor: Int) {
        let content: String
        if case .text(let t) = elem { content = t.content } else { content = "" }
        let s = TextEditSession(path: path, target: .text, content: content, insertion: cursor)
        s.blinkEpochMs = nowMs()
        session = s
        ctx.model.currentEditSession = s
        didSnapshot = false
        ctx.controller.selectElement(path)
    }

    private func beginSessionNew(_ ctx: ToolContext, x: Double, y: Double, w: Double, h: Double) {
        ctx.snapshot()
        didSnapshot = true
        let elem = Element.text(emptyTextElem(x: x, y: y, width: w, height: h))
        ctx.controller.addElement(elem)
        let doc = ctx.document
        let li = doc.selectedLayer
        let ci = doc.layers[li].children.count - 1
        let path: ElementPath = [li, ci]
        ctx.controller.selectElement(path)
        let s = TextEditSession(path: path, target: .text, content: "", insertion: 0)
        s.blinkEpochMs = nowMs()
        session = s
        ctx.model.currentEditSession = s
    }

    private func endSession(_ ctx: ToolContext? = nil) {
        session = nil
        didSnapshot = false
        state = .idle
        ctx?.model.currentEditSession = nil
    }

    func onPress(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, alt: Bool) {
        if let s = session {
            if pathIsValid(ctx.document, s.path) {
                let elem = ctx.document.getElement(s.path)
                let inElem = inBox(textDrawBounds(elem), x, y)
                if inElem {
                    let cursor = cursorAt(ctx, x, y)
                    s.setInsertion(cursor, extend: false)
                    s.dragActive = true
                    s.blinkEpochMs = nowMs()
                    ctx.requestUpdate()
                    return
                }
            }
            endSession(ctx)
        }
        if let (path, elem) = hitTestText(ctx.document, x, y) {
            beginSessionExisting(ctx, path: path, elem: elem, cursor: 0)
            if let s2 = session {
                let cursor = cursorAt(ctx, x, y)
                s2.setInsertion(cursor, extend: false)
                s2.dragActive = true
                s2.blinkEpochMs = nowMs()
                ctx.requestUpdate()
            }
        } else {
            state = .dragging(startX: x, startY: y, curX: x, curY: y)
        }
    }

    func onMove(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, dragging: Bool) {
        if let s = session, s.dragActive, dragging {
            let cursor = cursorAt(ctx, x, y)
            s.setInsertion(cursor, extend: true)
            s.blinkEpochMs = nowMs()
            ctx.requestUpdate()
            return
        }
        if case .dragging(let sx, let sy, _, _) = state {
            state = .dragging(startX: sx, startY: sy, curX: x, curY: y)
            ctx.requestUpdate()
        }
        if session == nil {
            hoverText = hitTestText(ctx.document, x, y) != nil
        } else {
            hoverText = false
        }
    }

    func onRelease(_ ctx: ToolContext, x: Double, y: Double, shift: Bool, alt: Bool) {
        if let s = session {
            s.dragActive = false
            s.blinkEpochMs = nowMs()
            state = .idle
            ctx.requestUpdate()
            return
        }
        guard case .dragging(let sx, let sy, _, _) = state else { return }
        state = .idle
        let w = abs(x - sx)
        let h = abs(y - sy)
        if w > dragThreshold || h > dragThreshold {
            beginSessionNew(ctx, x: min(sx, x), y: min(sy, y), w: w, h: h)
        } else {
            beginSessionNew(ctx, x: sx, y: sy, w: 0, h: 0)
        }
        ctx.requestUpdate()
    }

    func onDoubleClick(_ ctx: ToolContext, x: Double, y: Double) {
        guard let s = session else { return }
        s.selectAll()
        s.blinkEpochMs = nowMs()
        ctx.requestUpdate()
    }

    func deactivate(_ ctx: ToolContext) {
        endSession()
    }

    func capturesKeyboard() -> Bool { session != nil }
    func isEditing() -> Bool { session != nil }

    func cursorOverride() -> String? {
        // While an edit session is active or the pointer is hovering an
        // unlocked Text element, switch to the system I-beam.
        if session != nil { return "ibeam" }
        if hoverText { return "ibeam" }
        return nil
    }

    func pasteText(_ ctx: ToolContext, _ text: String) -> Bool {
        guard let s = session else { return false }
        // Rich-paste preference:
        //   1. Session-scoped tspan clipboard (within-session copy/paste).
        //   2. Rich-format pasteboard (application/x-jas-tspans, then
        //      image/svg+xml) — survives cross-app paste too.
        //   3. Flat insert.
        let elemTspans = currentElementTspans(ctx)
        var newTspans: [Tspan]? = s.tryPasteTspans(elemTspans, text: text)
        if newTspans == nil, let payload = richClipboardReadTspans() {
            newTspans = insertTspansAt(elemTspans, charPos: s.insertion, payload)
        }
        if let new = newTspans {
            ensureSnapshot(ctx)
            replaceElementTspans(ctx, path: s.path, tspans: new)
            s.setContent(concatTspanContent(new),
                         insertion: s.insertion + text.count,
                         anchor: s.insertion + text.count)
            s.blinkEpochMs = nowMs()
            ctx.requestUpdate()
            return true
        }
        ensureSnapshot(ctx)
        s.insert(text)
        s.blinkEpochMs = nowMs()
        syncToModel(ctx)
        ctx.requestUpdate()
        return true
    }

    /// Read the current element's tspans from the doc for cut/copy/paste.
    private func currentElementTspans(_ ctx: ToolContext) -> [Tspan] {
        guard let s = session else { return [] }
        let elem = ctx.model.document.getElement(s.path)
        switch elem {
        case .text(let t): return t.tspans
        case .textPath(let tp): return tp.tspans
        default: return []
        }
    }

    /// Replace the tspans on the element at `path` in the model's
    /// document. Used by the tspan-aware paste path.
    private func replaceElementTspans(_ ctx: ToolContext, path: ElementPath, tspans: [Tspan]) {
        let elem = ctx.model.document.getElement(path)
        let newElem: Element
        switch elem {
        case .text(let t): newElem = .text(t.withTspans(tspans))
        case .textPath(let tp): newElem = .textPath(tp.withTspans(tspans))
        default: return
        }
        ctx.model.document = ctx.model.document.replaceElement(path, with: newElem)
    }

    func onKeyEvent(_ ctx: ToolContext, _ key: String, _ mods: KeyMods) -> Bool {
        guard let s = session else { return false }
        let bump = { s.blinkEpochMs = nowMs() }
        let cmd = mods.cmd
        let lower = key.lowercased()

        if cmd && lower == "a" {
            s.selectAll(); bump(); ctx.requestUpdate(); return true
        }
        if cmd && lower == "z" {
            if mods.shift { s.redo() } else { s.undo() }
            bump(); syncToModel(ctx); ctx.requestUpdate(); return true
        }
        if cmd && lower == "c" {
            // Capture the selection's tspan overrides for both the
            // session-scoped fast path and the multi-format OS
            // pasteboard. Flat text goes as .string; the tspan list
            // rides on application/x-jas-tspans and image/svg+xml
            // for cross-element / cross-app paste.
            let elem = currentElementTspans(ctx)
            if let flat = s.copySelectionWithTspans(elem),
               let (_, payload) = s.tspanClipboard
            {
                richClipboardWrite(flat: flat, tspans: payload)
            }
            return true
        }
        if cmd && lower == "x" {
            let elem = currentElementTspans(ctx)
            if let flat = s.copySelectionWithTspans(elem),
               let (_, payload) = s.tspanClipboard
            {
                richClipboardWrite(flat: flat, tspans: payload)
                ensureSnapshot(ctx)
                s.backspace()
                bump(); syncToModel(ctx); ctx.requestUpdate()
            }
            return true
        }
        switch key {
        case "Escape":
            endSession(ctx); ctx.requestUpdate(); return true
        case "Enter":
            ensureSnapshot(ctx); s.insert("\n")
            bump(); syncToModel(ctx); ctx.requestUpdate(); return true
        case "Backspace":
            ensureSnapshot(ctx); s.backspace()
            bump(); syncToModel(ctx); ctx.requestUpdate(); return true
        case "Delete":
            ensureSnapshot(ctx); s.deleteForward()
            bump(); syncToModel(ctx); ctx.requestUpdate(); return true
        case "ArrowLeft":
            s.setInsertion(s.insertion - 1, extend: mods.shift)
            bump(); ctx.requestUpdate(); return true
        case "ArrowRight":
            s.setInsertion(s.insertion + 1, extend: mods.shift)
            bump(); ctx.requestUpdate(); return true
        case "ArrowUp", "ArrowDown":
            if let (_, lay) = buildLayout(ctx) {
                let np = key == "ArrowUp" ? lay.cursorUp(s.insertion) : lay.cursorDown(s.insertion)
                s.setInsertion(np, extend: mods.shift)
                bump(); ctx.requestUpdate()
            }
            return true
        case "Home":
            if let (_, lay) = buildLayout(ctx) {
                let lineNo = lay.lineForCursor(s.insertion)
                s.setInsertion(lay.lines[lineNo].start, extend: mods.shift)
                bump(); ctx.requestUpdate()
            }
            return true
        case "End":
            if let (_, lay) = buildLayout(ctx) {
                let lineNo = lay.lineForCursor(s.insertion)
                s.setInsertion(lay.lines[lineNo].end, extend: mods.shift)
                bump(); ctx.requestUpdate()
            }
            return true
        default:
            if key.count == 1 && !cmd {
                ensureSnapshot(ctx); s.insert(key)
                bump(); syncToModel(ctx); ctx.requestUpdate(); return true
            }
            return false
        }
    }

    func drawOverlay(_ ctx: ToolContext, _ cgCtx: CGContext) {
        // Drag-create rect preview
        if session == nil, case .dragging(let sx, let sy, let ex, let ey) = state {
            cgCtx.setStrokeColor(CGColor(gray: 0.4, alpha: 1.0))
            cgCtx.setLineWidth(1.0)
            cgCtx.setLineDash(phase: 0, lengths: [4, 4])
            let r = CGRect(x: min(sx, ex), y: min(sy, ey),
                           width: abs(ex - sx), height: abs(ey - sy))
            cgCtx.addRect(r)
            cgCtx.strokePath()
            cgCtx.setLineDash(phase: 0, lengths: [])
        }

        guard let s = session, let (tr, lay) = buildLayout(ctx) else { return }

        // Layout-local origin: `(x, y)` is the top of the layout box for
        // both point text and area text.
        let originX = tr.x
        let originY = tr.y

        // Selection highlight
        if s.hasSelection {
            let (lo, hi) = s.selectionRange
            cgCtx.setFillColor(CGColor(red: 0.529, green: 0.808, blue: 0.980, alpha: 0.45))
            for (lineIdx, line) in lay.lines.enumerated() {
                let lineLo = max(line.start, lo)
                let lineHi = min(line.end, hi)
                if lineLo >= lineHi { continue }
                func glyphX(_ i: Int) -> Double? {
                    for g in lay.glyphs where g.idx == i && g.line == lineIdx { return g.x }
                    return nil
                }
                let xLo = lineLo == line.start ? 0.0 : (glyphX(lineLo) ?? 0.0)
                let xHi = lineHi == line.end ? line.width : (glyphX(lineHi) ?? line.width)
                let r = CGRect(x: originX + xLo, y: originY + line.top,
                               width: xHi - xLo, height: line.height)
                cgCtx.fill(r)
            }
        }

        // The bounding box around the edited text is not drawn here —
        // the Type tool selects the element when it starts editing,
        // so the selection overlay (see `drawElementOverlay` /
        // `drawSelectionOverlays` in `Canvas/CanvasSubwindow.swift`)
        // is responsible for rendering the box. That keeps the rule
        // "area text shows its bbox iff the element is selected" in
        // a single place.

        // Caret
        if cursorVisible(s.blinkEpochMs) {
            let (cx, cy, ch) = lay.cursorXY(s.insertion)
            let color: Color
            if let f = tr.fill { color = f.color }
            else if let st = tr.stroke { color = st.color }
            else { color = Color(r: 0, g: 0, b: 0) }
            let (cr, cg, cb, _) = color.toRgba()
            cgCtx.setStrokeColor(CGColor(red: cr, green: cg, blue: cb, alpha: 1.0))
            cgCtx.setLineWidth(1.5)
            cgCtx.move(to: CGPoint(x: originX + cx, y: originY + cy - ch * 0.8))
            cgCtx.addLine(to: CGPoint(x: originX + cx, y: originY + cy + ch * 0.2))
            cgCtx.strokePath()
        }
    }
}

File: jas/tools/type_tool.py
"""Type tool with native in-place text editing.

Click on existing unlocked text to edit it; click on empty canvas to
create a new (initially empty) text element and immediately enter editing
mode. Drag to create an area text box.

While editing:
 - Mouse drag inside the editing element extends the selection.
 - Standard text editing keys (arrows, backspace, delete, Home/End,
   Cmd+A/C/X/V/Z) are routed to the session via `on_key_event`.
 - The OS cursor is hidden over the active edit area; a vertical text
   caret is drawn at the insertion point.
 - All character-level edits go through a per-session undo stack and
   collapse to a *single* document-undo step.

See `jas_dioxus/src/tools/text_edit.rs` for the full design notes.
"""

from __future__ import annotations

import dataclasses
from typing import TYPE_CHECKING

from geometry.element import (
    Color, Element, Fill, Group, Layer, Text, Stroke,
)
from algorithms.text_layout import layout as _layout, TextLayout
from tools.text_edit import (
    EditTarget, TextEditSession, empty_text_elem,
    BLINK_HALF_PERIOD_MS, now_ms as _now_ms, cursor_visible as _cursor_visible,
)
from tools.text_measure import make_measurer
from tools.tool import CanvasTool, KeyMods, ToolContext, DRAG_THRESHOLD

if TYPE_CHECKING:
    from PySide6.QtGui import QPainter


def _show_selection_bbox() -> bool:
    """Indirection to canvas.SHOW_SELECTION_BBOX so the type tool can
    consult it without a hard import-time dependency on the canvas
    module (which transitively imports Qt)."""
    from canvas.canvas import SHOW_SELECTION_BBOX
    return SHOW_SELECTION_BBOX


def _text_draw_bounds(t: Text) -> tuple[float, float, float, float]:
    """Bounding box used to hit-test a Text element. Both point and area
    text are treated as having `e.y` at their top edge."""
    if t.is_area_text:
        return (t.x, t.y, max(t.width, 1.0), max(t.height, 1.0))
    lines = t.content.split('\n') if t.content else [""]
    max_chars = max(len(l) for l in lines) if lines else 0
    w = max(max_chars, 1) * t.font_size * 0.55
    h = len(lines) * t.font_size
    return (t.x, t.y, w, h)


def _accent_color(t: Text) -> "QColor":
    from PySide6.QtGui import QColor
    c = (t.fill.color if t.fill is not None
         else (t.stroke.color if t.stroke is not None else Color(0, 0, 0)))
    return QColor(int(c.r * 255), int(c.g * 255), int(c.b * 255))


def _selection_color(t: Text) -> "QColor":
    """Light sky blue, falling back to yellow if too close to fill or stroke."""
    from PySide6.QtGui import QColor
    blue_lum = 0.2126 * 0.529 + 0.7152 * 0.808 + 0.0722 * 0.980
    candidates = []
    if t.fill is not None:
        candidates.append(t.fill.color)
    if t.stroke is not None:
        candidates.append(t.stroke.color)
    for c in candidates:
        lum = 0.2126 * c.r + 0.7152 * c.g + 0.0722 * c.b
        if abs(lum - blue_lum) < 0.15:
            return QColor(255, 235, 80, 128)
    return QColor(135, 206, 250, 115)


@dataclasses.dataclass
class _Dragging:
    start_x: float
    start_y: float
    cur_x: float
    cur_y: float


class TypeTool(CanvasTool):
    def __init__(self):
        # Drag-to-create state machine. None == idle.
        self._drag: _Dragging | None = None
        # active edit session, if any
        self.session: TextEditSession | None = None
        self._did_snapshot = False
        self._hover_text = False

    # ---- helpers ----

    def _build_layout(self, ctx: ToolContext) -> tuple[Text, TextLayout] | None:
        if self.session is None or self.session.target != EditTarget.TEXT:
            return None
        try:
            elem = ctx.document.get_element(self.session.path)
        except (IndexError, KeyError):
            return None
        if not isinstance(elem, Text):
            return None
        t = dataclasses.replace(elem, content=self.session.content)
        measure = make_measurer(t.font_family, t.font_weight, t.font_style, t.font_size)
        max_w = t.width if t.is_area_text else 0.0
        lay = _layout(t.content, max_w, t.font_size, measure)
        return (t, lay)

    def _hit_test_text(self, ctx: ToolContext, x: float, y: float
                       ) -> tuple[tuple[int, ...], Text] | None:
        """Recursive hit-test that respects locked groups/elements."""
        doc = ctx.document
        result: list[tuple[tuple[int, ...], Text] | None] = [None]

        def rec(elem, path):
            if isinstance(elem, Layer):
                for i, c in enumerate(elem.children):
                    rec(c, path + (i,))
            elif isinstance(elem, Group):
                if elem.locked:
                    return
                for i, c in enumerate(elem.children):
                    rec(c, path + (i,))
            elif isinstance(elem, Text):
                if elem.locked:
                    return
                bx, by, bw, bh = _text_draw_bounds(elem)
                if bx <= x <= bx + bw and by <= y <= by + bh:
                    result[0] = (path, elem)
        for li, layer in enumerate(doc.layers):
            rec(layer, (li,))
        return result[0]

    def _cursor_at(self, ctx: ToolContext, x: float, y: float) -> int:
        built = self._build_layout(ctx)
        if built is None:
            return 0
        t, lay = built
        return lay.hit_test(x - t.x, y - t.y)

    def _ensure_snapshot(self, ctx: ToolContext) -> None:
        if not self._did_snapshot:
            ctx.snapshot()
            self._did_snapshot = True

    def _sync_to_model(self, ctx: ToolContext) -> None:
        if self.session is None:
            return
        new_doc = self.session.apply_to_document(ctx.document)
        if new_doc is not None:
            ctx.controller.set_document(new_doc)

    def _begin_session_existing(self, ctx: ToolContext, path: tuple,
                                elem: Text, cursor: int) -> None:
        self.session = TextEditSession(
            path=path, target=EditTarget.TEXT,
            content=elem.content, insertion=cursor,
            blink_epoch_ms=_now_ms(),
        )
        self._did_snapshot = False
        ctx.controller.select_element(path)

    def _begin_session_new(self, ctx: ToolContext, x: float, y: float,
                           width: float, height: float) -> None:
        ctx.snapshot()
        self._did_snapshot = True
        elem = empty_text_elem(x, y, width, height)
        ctx.controller.add_element(elem)
        doc = ctx.document
        li = doc.selected_layer
        ci = len(doc.layers[li].children) - 1
        path = (li, ci)
        ctx.controller.select_element(path)
        self.session = TextEditSession(
            path=path, target=EditTarget.TEXT, content="",
            insertion=0, blink_epoch_ms=_now_ms(),
        )

    def _end_session(self) -> None:
        self.session = None
        self._did_snapshot = False
        self._drag = None

    # ---- CanvasTool API ----

    def on_press(self, ctx: ToolContext, x: float, y: float,
                 shift: bool = False, alt: bool = False) -> None:
        if self.session is not None:
            try:
                elem = ctx.document.get_element(self.session.path)
            except (IndexError, KeyError):
                elem = None
            in_elem = False
            if isinstance(elem, Text):
                bx, by, bw, bh = _text_draw_bounds(elem)
                in_elem = bx <= x <= bx + bw and by <= y <= by + bh
            if in_elem:
                cursor = self._cursor_at(ctx, x, y)
                self.session.set_insertion(cursor, False)
                self.session.drag_active = True
                self.session.blink_epoch_ms = _now_ms()
                ctx.request_update()
                return
            self._end_session()

        hit = self._hit_test_text(ctx, x, y)
        if hit is not None:
            path, t = hit
            self._begin_session_existing(ctx, path, t, 0)
            cursor = self._cursor_at(ctx, x, y)
            self.session.set_insertion(cursor, False)
            self.session.drag_active = True
            self.session.blink_epoch_ms = _now_ms()
            ctx.request_update()
            return

        self._drag = _Dragging(start_x=x, start_y=y, cur_x=x, cur_y=y)

    def on_move(self, ctx: ToolContext, x: float, y: float,
                shift: bool = False, dragging: bool = False) -> None:
        if self.session is not None and self.session.drag_active and dragging:
            cursor = self._cursor_at(ctx, x, y)
            self.session.set_insertion(cursor, True)
            self.session.blink_epoch_ms = _now_ms()
            ctx.request_update()
            return
        if self._drag is not None:
            self._drag.cur_x = x
            self._drag.cur_y = y
            ctx.request_update()
        if self.session is None:
            self._hover_text = self._hit_test_text(ctx, x, y) is not None
        else:
            self._hover_text = False

    def on_release(self, ctx: ToolContext, x: float, y: float,
                   shift: bool = False, alt: bool = False) -> None:
        if self.session is not None:
            self.session.drag_active = False
            self.session.blink_epoch_ms = _now_ms()
            self._drag = None
            ctx.request_update()
            return
        if self._drag is None:
            return
        sx, sy = self._drag.start_x, self._drag.start_y
        self._drag = None
        w = abs(x - sx)
        h = abs(y - sy)
        if w > DRAG_THRESHOLD or h > DRAG_THRESHOLD:
            bx, by = min(sx, x), min(sy, y)
            self._begin_session_new(ctx, bx, by, w, h)
        else:
            self._begin_session_new(ctx, sx, sy, 0.0, 0.0)
        ctx.request_update()

    def on_double_click(self, ctx: ToolContext, x: float, y: float) -> None:
        if self.session is not None:
            self.session.select_all()
            self.session.blink_epoch_ms = _now_ms()
            ctx.request_update()

    def on_key_event(self, ctx: ToolContext, key: str, mods: KeyMods) -> bool:
        if self.session is None:
            return False

        if mods.cmd():
            if key in ("a", "A"):
                self.session.select_all()
                self.session.blink_epoch_ms = _now_ms()
                ctx.request_update()
                return True
            if key in ("z", "Z"):
                if mods.shift:
                    self.session.redo()
                else:
                    self.session.undo()
                self.session.blink_epoch_ms = _now_ms()
                self._sync_to_model(ctx)
                ctx.request_update()
                return True
            if key in ("c", "C"):
                text = self.session.copy_selection()
                if text is not None:
                    _clipboard_write(text)
                return True
            if key in ("x", "X"):
                text = self.session.copy_selection()
                if text is not None:
                    _clipboard_write(text)
                    self._ensure_snapshot(ctx)
                    self.session.backspace()
                    self.session.blink_epoch_ms = _now_ms()
                    self._sync_to_model(ctx)
                    ctx.request_update()
                return True

        if key == "Escape":
            self._end_session()
            ctx.request_update()
            return True
        if key == "Enter":
            self._ensure_snapshot(ctx)
            self.session.insert("\n")
            self.session.blink_epoch_ms = _now_ms()
            self._sync_to_model(ctx)
            ctx.request_update()
            return True
        if key == "Backspace":
            self._ensure_snapshot(ctx)
            self.session.backspace()
            self.session.blink_epoch_ms = _now_ms()
            self._sync_to_model(ctx)
            ctx.request_update()
            return True
        if key == "Delete":
            self._ensure_snapshot(ctx)
            self.session.delete_forward()
            self.session.blink_epoch_ms = _now_ms()
            self._sync_to_model(ctx)
            ctx.request_update()
            return True
        if key == "ArrowLeft":
            self.session.set_insertion(max(0, self.session.insertion - 1), mods.shift)
            self.session.blink_epoch_ms = _now_ms()
            ctx.request_update()
            return True
        if key == "ArrowRight":
            self.session.set_insertion(self.session.insertion + 1, mods.shift)
            self.session.blink_epoch_ms = _now_ms()
            ctx.request_update()
            return True
        if key in ("ArrowUp", "ArrowDown"):
            built = self._build_layout(ctx)
            if built is not None:
                _, lay = built
                if key == "ArrowUp":
                    new_pos = lay.cursor_up(self.session.insertion)
                else:
                    new_pos = lay.cursor_down(self.session.insertion)
                self.session.set_insertion(new_pos, mods.shift)
                self.session.blink_epoch_ms = _now_ms()
                ctx.request_update()
            return True
        if key == "Home":
            built = self._build_layout(ctx)
            if built is not None:
                _, lay = built
                line = lay.line_for_cursor(self.session.insertion)
                self.session.set_insertion(lay.lines[line].start, mods.shift)
                self.session.blink_epoch_ms = _now_ms()
                ctx.request_update()
            return True
        if key == "End":
            built = self._build_layout(ctx)
            if built is not None:
                _, lay = built
                line = lay.line_for_cursor(self.session.insertion)
                self.session.set_insertion(lay.lines[line].end, mods.shift)
                self.session.blink_epoch_ms = _now_ms()
                ctx.request_update()
            return True

        if len(key) == 1 and not mods.cmd():
            self._ensure_snapshot(ctx)
            self.session.insert(key)
            self.session.blink_epoch_ms = _now_ms()
            self._sync_to_model(ctx)
            ctx.request_update()
            return True
        return False

    def captures_keyboard(self) -> bool:
        return self.session is not None

    def cursor_css_override(self) -> str | None:
        # While editing, always use the system I-beam.
        if self.session is not None:
            return "ibeam"
        if self._hover_text:
            return "ibeam"
        return None

    def is_editing(self) -> bool:
        return self.session is not None

    def paste_text(self, ctx: ToolContext, text: str) -> bool:
        if self.session is None:
            return False
        self._ensure_snapshot(ctx)
        self.session.insert(text)
        self.session.blink_epoch_ms = _now_ms()
        self._sync_to_model(ctx)
        ctx.request_update()
        return True

    def deactivate(self, ctx: ToolContext) -> None:
        self._end_session()

    def draw_overlay(self, ctx: ToolContext, painter: "QPainter") -> None:
        from PySide6.QtCore import QPointF, QRectF, Qt
        from PySide6.QtGui import QBrush, QColor, QPen

        # Drag-create preview rectangle.
        if self.session is None and self._drag is not None:
            sx, sy = self._drag.start_x, self._drag.start_y
            ex, ey = self._drag.cur_x, self._drag.cur_y
            pen = QPen(QColor(100, 100, 100), 1.0, Qt.PenStyle.DashLine)
            painter.setPen(pen)
            painter.setBrush(QBrush())
            painter.drawRect(QRectF(QPointF(sx, sy), QPointF(ex, ey)).normalized())

        if self.session is None:
            return
        built = self._build_layout(ctx)
        if built is None:
            return
        t, lay = built

        sel_color = _selection_color(t)
        caret_color = _accent_color(t)

        # Selection rectangles.
        if self.session.has_selection():
            lo, hi = self.session.selection_range()
            painter.setPen(Qt.PenStyle.NoPen)
            painter.setBrush(QBrush(sel_color))
            for line_idx, line in enumerate(lay.lines):
                line_lo = max(line.start, lo)
                line_hi = min(line.end, hi)
                if line_lo >= line_hi:
                    continue
                if line_lo == line.start:
                    x_lo = 0.0
                else:
                    x_lo = next((g.x for g in lay.glyphs
                                 if g.idx == line_lo and g.line == line_idx), 0.0)
                if line_hi == line.end:
                    x_hi = line.width
                else:
                    x_hi = next((g.x for g in lay.glyphs
                                 if g.idx == line_hi and g.line == line_idx), line.width)
                painter.drawRect(QRectF(t.x + x_lo, t.y + line.top,
                                        x_hi - x_lo, line.height))

        # The bounding box around the edited text is not drawn
        # here — the Type tool selects the element when it starts
        # editing, so the selection overlay (see
        # ``_draw_selection_overlays`` in ``canvas/canvas.py``) is
        # responsible for rendering the box. That keeps the rule
        # "area text shows its bbox iff the element is selected"
        # in a single place.

        # Caret.
        if _cursor_visible(self.session.blink_epoch_ms):
            cx, cy, ch = lay.cursor_xy(self.session.insertion)
            painter.setPen(QPen(caret_color, 1.5))
            painter.drawLine(QPointF(t.x + cx, t.y + cy - ch * 0.8),
                             QPointF(t.x + cx, t.y + cy + ch * 0.2))


def _clipboard_write(text: str) -> None:
    try:
        from PySide6.QtWidgets import QApplication  # type: ignore
        app = QApplication.instance()
        if app is not None:
            app.clipboard().setText(text)
    except (ImportError, AttributeError):
        pass
